src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/SelectExpandNode.cs

```diff
@@ -242,30 +253,125 @@ public SelectExpandNode(SelectExpandClause selectExpandClause, IEdmStructuredTyp
-            foreach (SelectItem selectItem in selectExpandClause.SelectedItems)
+            foreach (SelectItem selectItem in selectedItems)
             {
                 ExpandedReferenceSelectItem expandReferenceItem = selectItem as ExpandedReferenceSelectItem;
                 if (expandReferenceItem != null)
                 {
-                    ValidatePathIsSupported(expandReferenceItem.PathToNavigationProperty);
+                    ValidatePathIsSupportedForExpand(expandReferenceItem.PathToNavigationProperty);
                     NavigationPropertySegment navigationSegment = (NavigationPropertySegment)expandReferenceItem.PathToNavigationProperty.LastSegment;
                     IEdmNavigationProperty navigationProperty = navigationSegment.NavigationProperty;
                     if (allNavigationProperties.Contains(navigationProperty))
                     {
                         ExpandedNavigationSelectItem expandItem = selectItem as ExpandedNavigationSelectItem;
+
                         if (expandItem != null)
                         {
-                            ExpandedProperties.Add(navigationProperty, expandItem);
+                            if (!ExpandedProperties.ContainsKey(navigationProperty))
+                            {
+                                ExpandedProperties.Add(navigationProperty, expandItem);
+                            }
+                            else
+                            {
+                                ExpandedProperties[navigationProperty] = expandItem;
+                            }
                         }
                         else
                         {
                             ReferencedNavigationProperties.Add(navigationProperty);
                         }
                     }
+                    else
+                    {
+                        //This is the case where the navigation property is not on the current type. We need to propagate the expand item to deeper SelectExpandNode.
+                        IEdmProperty complexProperty = FindNextPropertySegment(expandReferenceItem.PathToNavigationProperty);
+
+                        if (complexProperty != null && ExpandedNavigationPropertiesOnComplexTypes.ContainsKey(complexProperty))
+                        {
+                            SelectExpandClause oldClause = ExpandedNavigationPropertiesOnComplexTypes[complexProperty].SelectAndExpand;
+                            SelectExpandClause newClause = new SelectExpandClause(oldClause.SelectedItems.Concat(new List<SelectItem>() { expandReferenceItem }), false);
+                            ExpandedNavigationSelectItem newItem = new ExpandedNavigationSelectItem(expandReferenceItem.PathToNavigationProperty, navigationSegment.NavigationSource, newClause);
+                            ExpandedNavigationPropertiesOnComplexTypes.Add(complexProperty, newItem);
+                        }
+                        else if (complexProperty != null)
```

Comment:
wouldn't it be better to do if(complexProperty !=null) on 301, and then nest if(ExpandedNavigationPropertiesOnComplexType.ContainsKey() .. else) so that you don't have to check complexProperty != null twice?